src/Components/Components/src/Routing/OptionalTypeRouteConstraint.cs

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Components.Routing
+{
+    /// <summary>
+    /// A route constraint that allows the value to be null or parseable as the specified
+    /// type.
+    /// </summary>
+    /// <typeparam name="T">The type to which the value must be parseable.</typeparam>
+    internal class OptionalTypeRouteConstraint<T> : RouteConstraint
+    {
+        public delegate bool TryParseDelegate(string str, out T result);
+
+        private readonly TryParseDelegate _parser;
+
+        public OptionalTypeRouteConstraint(TryParseDelegate parser)
+        {
+            _parser = parser;
+        }
+
+        public override bool Match(string pathSegment, out object convertedValue)
+        {
+            if (string.IsNullOrEmpty(pathSegment))
+            {
+                convertedValue = null;
```

Comment:
Is it safe to use `null` here instead of `default(T)`? Is the idea that a layer about this reinterprets null based on the property it's getting assigned to? Might be worth a comment

Reply:
Yep. The parameters object that is populated in the RouteContext expects unset values to be set to `null`. I can add a comment about this.